dnf5-plugins/reposync_plugin/reposync.cpp

```diff
@@ -0,0 +1,420 @@
+    if (ec) {
+        std::cerr << libdnf5::utils::sformat(
+                         _("Failed to create directory '{0}' iterator: {1}"), repo_path.string(), ec.message())
+                  << std::endl;
+        return;
+    }
+
+    std::unordered_set<std::string> downloaded_paths;
+    for (const auto & [pth, pkg] : pkg_list) {
+        downloaded_paths.emplace(pth);
+    }
+
+    for (const auto & entry : delete_iterator) {
+        if (entry.is_regular_file(ec)) {
+            const auto & file_path = entry.path();
+            if (file_path.extension() == ".rpm" && !downloaded_paths.contains(file_path)) {
+                // Remove every *.rpm file that was not downloaded from the repo
+                std::filesystem::remove(file_path, ec);
+                if (ec) {
+                    std::cerr << libdnf5::utils::sformat(
+                                     _("Failed to delete file {0}: {1}"), file_path.string(), ec.message())
+                              << std::endl;
+                } else {
+                    std::cout << libdnf5::utils::sformat(_("[DELETED] {}"), file_path.string()) << std::endl;
+                }
+            }
+        }
+    }
+}
+
+bool ReposyncCommand::pgp_check_packages(const download_list_type & pkg_list) {
+    bool ret = true;
+    std::error_code ec;
+    libdnf5::rpm::RpmSignature rpm_signature(get_context().get_base());
+    for (const auto & [pth, pkg] : pkg_list) {
+        if (std::filesystem::exists(pth, ec)) {
+            auto check_result = rpm_signature.check_package_signature(pth);
+            if (check_result != libdnf5::rpm::RpmSignature::CheckResult::OK) {
+                std::cerr << libdnf5::utils::sformat(
+                                 _("Removing '{}' with failing PGP check: {}"),
+                                 pth.string(),
+                                 rpm_signature.check_result_to_string(check_result))
+                          << std::endl;
+                std::filesystem::remove(pth, ec);
+                ret = false;
+            }
+        }
+    }
+    return ret;
+}
+
+void ReposyncCommand::download_metadata(libdnf5::repo::Repo & repo) {
+    std::filesystem::path repo_path;
+    const auto metadata_path = metadata_path_option->get_value();
+    if (!metadata_path.empty()) {
+        repo_path = std::filesystem::absolute(metadata_path);
+        if (!norepopath_option->get_value()) {
+            repo_path /= repo.get_id();
+        }
+        // resolve '.', '..', and existing symlinks in the repo_path
+        repo_path = std::filesystem::weakly_canonical(repo_path);
+    } else {
+        repo_path = repo_download_path(repo);
+    }
+    auto & optional_metadata_option = get_context().get_base().get_config().get_optional_metadata_types_option();
+    if (!optional_metadata_option.get_value().contains(libdnf5::METADATA_TYPE_ALL)) {
+        optional_metadata_option.set(libdnf5::METADATA_TYPE_ALL);
+    }
+    repo.download_metadata(repo_path);
```

Comment:
If I understand correctly we download the same metadata for the second time here. All enabled repositories were `expired` in the `configure` step so we could just copy those and download only the missing parts.

I think this would be a nice optimization but I wouldn't block this PR on that.